### DISCUSSION TITLE
Calibration

### DISCUSSION THREAD

**Discussion by bormilan** [Kalamansanai/plm-ssp#59] [Backend]
# Calibration sequence from the backend side

## Endpoints
* `Identify`
    * This is the identifying method of the Detector, it gets a `locationId`, a `mac`, and the `coordinates` of the QR code.
    * If there is no detector with the given `mac`, it creates a new, and adds it to the location.
    * If the detector is available, and the location has no already connected with a detector, adds it to the location
    * If the detector had no `originalCoordinates` before, it saves the `newCoordinates` into it
    * If the new coordinates are not the same as the originals, calls the detector entity's `sendRecalibrate` function
* `ReCalibrate`
    * This is for the manual calibration, the frontend calls it with a `locationId`, and it sends the `originalCoordinates` of the detector to the RPI.

## Methods
* The Detector entity has a function called `SendRecalibrate`, it sends a http request to the detector with the `originalCoordinates`.

## Questions
* If we recalibrated a detector, and wants to recalibrate again, we need to give the first original coordinates, or the new ?
* When we recalibrate by hand, the new coordinates will not be stored, but during auto calibration they are. Is it a good logic ? (i think its weird, and not gona be good )

**Comment by bormilan** [comment]
@Fullbaro @anguaive

**Comment by sij1nk** [comment]
You'll need to implement the frontend-facing API as described in https://github.com/Kalamansanai/plm-ssp/wiki/Calibration if you want to keep all the previous functionality

> If we recalibrated a detector, and wants to recalibrate again, we need to give the first original coordinates, or the new ?

You only keep the last set of tray keypoints, either sent by the frontend (manual calibration) or the RPI (in case of automatic calibration, as seen here: https://github.com/bmetk/plm/blob/master/backend-dotnet-new/Services/DeviceControlHandler.cs#L278)

> When we recalibrate by hand, the new coordinates will not be stored, but during auto calibration they are. Is it a good logic ? (i think its weird, and not gona be good )

No, applying the manual calibration (user sets the tray kps and clicks apply -> kps are saved) and previewing it (user sets the tray kps and clicks preview to see how the calibrated output would look like -> kps are not saved) should be handled separately. 

The reason we don't save the keypoints on preview is:
 - the user should be able to see what the calibration result would look like before applying it 
 - if it takes the user 10 preview requests to configure the keypoints, we only really care about the last set (when they click apply)
